File: matrix-1.py
import random
import time
import math

# BATCH_SIZE = 1000
# INPUT_SIZE = 784
# HIDDEN_SIZE = 512
# OUTPUT_SIZE = 10
# ITERATIONS = 50

BATCH_SIZE = 1000
INPUT_SIZE = 78
HIDDEN_SIZE = 51
OUTPUT_SIZE = 10
ITERATIONS = 50

def create_matrix(rows, cols, fill=None):
    if fill is None:
        return [[random.gauss(0, 1) for _ in range(cols)] for _ in range(rows)]
    return [[fill for _ in range(cols)] for _ in range(rows)]

def create_vector(size, fill=None):
    if fill is None:
        return [random.gauss(0, 1) for _ in range(size)]
    return [fill for _ in range(size)]

def matrix_multiply(a, b):
    rows_a = len(a)
    cols_a = len(a[0])
    cols_b = len(b[0])
    result = [[0 for _ in range(cols_b)] for _ in range(rows_a)]

    for i in range(rows_a):
        for j in range(cols_b):
            for k in range(cols_a):
                result[i][j] += a[i][k] * b[k][j]
    return result

def transpose(matrix):
    return [[matrix[j][i] for j in range(len(matrix))] for i in range(len(matrix[0]))]

def add_matrix(a, b):
    return [[a[i][j] + b[i][j] for j in range(len(a[0]))] for i in range(len(a))]

def subtract_matrix(a, b):
    return [[a[i][j] - b[i][j] for j in range(len(a[0]))] for i in range(len(a))]

def add_vector_to_matrix(matrix, vector):
    return [[matrix[i][j] + vector[j] for j in range(len(matrix[0]))] for i in range(len(matrix))]

def relu(matrix):
    return [[max(0, val) for val in row] for row in matrix]

def relu_derivative(matrix):
    return [[1 if val > 0 else 0 for val in row] for row in matrix]

def softmax(matrix):
    result = [[0 for _ in range(len(matrix[0]))] for _ in range(len(matrix))]
    for i in range(len(matrix)):
        row_max = max(matrix[i])
        exp_sum = sum(math.exp(x - row_max) for x in matrix[i])
        for j in range(len(matrix[0])):
            result[i][j] = math.exp(matrix[i][j] - row_max) / exp_sum
    return result

def hadamard_product(a, b):
    return [[a[i][j] * b[i][j] for j in range(len(a[0]))] for i in range(len(a))]

def sum_columns(matrix):
    return [sum(row[j] for row in matrix) for j in range(len(matrix[0]))]

def generate_data():
    X = create_matrix(BATCH_SIZE, INPUT_SIZE)
    y = [[0 for _ in range(OUTPUT_SIZE)] for _ in range(BATCH_SIZE)]
    for i in range(BATCH_SIZE):
        target = random.randint(0, OUTPUT_SIZE - 1)
        y[i][target] = 1
    return X, y

def forward(X, W1, b1, W2, b2):
    z1 = add_vector_to_matrix(matrix_multiply(X, W1), b1)
    a1 = relu(z1)
    z2 = add_vector_to_matrix(matrix_multiply(a1, W2), b2)
    probs = softmax(z2)
    return z1, a1, z2, probs

def backward(X, y, probs, z1, a1, W1, W2):
    dz2 = subtract_matrix(probs, y)
    dW2 = matrix_multiply(transpose(a1), dz2)
    db2 = sum_columns(dz2)

    da1 = matrix_multiply(dz2, transpose(W2))
    dz1 = hadamard_product(da1, relu_derivative(z1))
    dW1 = matrix_multiply(transpose(X), dz1)
    db1 = sum_columns(dz1)

    return dW1, db1, dW2, db2

def init_weights():
    scale1 = math.sqrt(2.0/INPUT_SIZE)
    scale2 = math.sqrt(2.0/HIDDEN_SIZE)
    W1 = [[random.gauss(0, 1) * scale1 for _ in range(HIDDEN_SIZE)] for _ in range(INPUT_SIZE)]
    b1 = create_vector(HIDDEN_SIZE, 0)
    W2 = [[random.gauss(0, 1) * scale2 for _ in range(OUTPUT_SIZE)] for _ in range(HIDDEN_SIZE)]
    b2 = create_vector(OUTPUT_SIZE, 0)
    return W1, b1, W2, b2

def main():
    W1, b1, W2, b2 = init_weights()
    X, y = generate_data()

    forward_time = 0
    backward_time = 0

    for _ in range(ITERATIONS):
        start = time.perf_counter()
        z1, a1, z2, probs = forward(X, W1, b1, W2, b2)
        forward_time += time.perf_counter() - start

        start = time.perf_counter()
        dW1, db1, dW2, db2 = backward(X, y, probs, z1, a1, W1, W2)
        backward_time += time.perf_counter() - start

    print(f'Размеры матриц:')
    print(f'Батч: {BATCH_SIZE}')
    print(f'Вход: {INPUT_SIZE}')
    print(f'Скрытый слой: {HIDDEN_SIZE}')
    print(f'Выход: {OUTPUT_SIZE}')
    print(f'\nВремя выполнения ({ITERATIONS} итераций):')
    print(f'Forward pass: {forward_time/ITERATIONS:.6f} сек')
    print(f'Backward pass: {backward_time/ITERATIONS:.6f} сек')
    print(f'Общее время: {(forward_time + backward_time)/ITERATIONS:.6f} сек')

if __name__ == '__main__':
    main()

File: matrix-1.js
// const BATCH_SIZE = 1000;
// const INPUT_SIZE = 784;
// const HIDDEN_SIZE = 512;
// const OUTPUT_SIZE = 10;
// const ITERATIONS = 50;

const BATCH_SIZE = 1000;
const INPUT_SIZE = 78;
const HIDDEN_SIZE = 51;
const OUTPUT_SIZE = 10;
const ITERATIONS = 50;

function createMatrix(rows, cols, fill = null) {
  if (fill === null) {
    return Array(rows)
      .fill()
      .map(() =>
        Array(cols)
          .fill()
          .map(() => randn())
      );
  }
  return Array(rows)
    .fill()
    .map(() => Array(cols).fill(fill));
}

function createVector(size, fill = null) {
  if (fill === null) {
    return Array(size)
      .fill()
      .map(() => randn());
  }
  return Array(size).fill(fill);
}

function randn() {
  let u = 0,
    v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  return Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
}

function matrixMultiply(a, b) {
  const rows = a.length;
  const cols = b[0].length;
  const result = Array(rows)
    .fill()
    .map(() => Array(cols).fill(0));

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      for (let k = 0; k < b.length; k++) {
        result[i][j] += a[i][k] * b[k][j];
      }
    }
  }
  return result;
}

function transpose(matrix) {
  return matrix[0].map((_, i) => matrix.map((row) => row[i]));
}

function addMatrix(a, b) {
  return a.map((row, i) => row.map((val, j) => val + b[i][j]));
}

function subtractMatrix(a, b) {
  return a.map((row, i) => row.map((val, j) => val - b[i][j]));
}

function addVectorToMatrix(matrix, vector) {
  return matrix.map((row) => row.map((val, j) => val + vector[j]));
}

function relu(matrix) {
  return matrix.map((row) => row.map((val) => Math.max(0, val)));
}

function reluDerivative(matrix) {
  return matrix.map((row) => row.map((val) => (val > 0 ? 1 : 0)));
}

function softmax(matrix) {
  return matrix.map((row) => {
    const rowMax = Math.max(...row);
    const expValues = row.map((val) => Math.exp(val - rowMax));
    const expSum = expValues.reduce((a, b) => a + b);
    return expValues.map((val) => val / expSum);
  });
}

function hadamardProduct(a, b) {
  return a.map((row, i) => row.map((val, j) => val * b[i][j]));
}

function sumColumns(matrix) {
  return matrix[0].map((_, j) => matrix.reduce((sum, row) => sum + row[j], 0));
}

function generateData() {
  const X = createMatrix(BATCH_SIZE, INPUT_SIZE);
  const y = createMatrix(BATCH_SIZE, OUTPUT_SIZE, 0);
  for (let i = 0; i < BATCH_SIZE; i++) {
    const target = Math.floor(Math.random() * OUTPUT_SIZE);
    y[i][target] = 1;
  }
  return [X, y];
}

function forward(X, W1, b1, W2, b2) {
  const z1 = addVectorToMatrix(matrixMultiply(X, W1), b1);
  const a1 = relu(z1);
  const z2 = addVectorToMatrix(matrixMultiply(a1, W2), b2);
  const probs = softmax(z2);
  return [z1, a1, z2, probs];
}

function backward(X, y, probs, z1, a1, W1, W2) {
  const dz2 = subtractMatrix(probs, y);
  const dW2 = matrixMultiply(transpose(a1), dz2);
  const db2 = sumColumns(dz2);

  const da1 = matrixMultiply(dz2, transpose(W2));
  const dz1 = hadamardProduct(da1, reluDerivative(z1));
  const dW1 = matrixMultiply(transpose(X), dz1);
  const db1 = sumColumns(dz1);

  return [dW1, db1, dW2, db2];
}

function initWeights() {
  const scale1 = Math.sqrt(2.0 / INPUT_SIZE);
  const scale2 = Math.sqrt(2.0 / HIDDEN_SIZE);

  const W1 = createMatrix(INPUT_SIZE, HIDDEN_SIZE).map((row) => row.map((val) => val * scale1));
  const b1 = createVector(HIDDEN_SIZE, 0);
  const W2 = createMatrix(HIDDEN_SIZE, OUTPUT_SIZE).map((row) => row.map((val) => val * scale2));
  const b2 = createVector(OUTPUT_SIZE, 0);

  return [W1, b1, W2, b2];
}

function main() {
  const [W1, b1, W2, b2] = initWeights();
  const [X, y] = generateData();

  let forwardTime = 0;
  let backwardTime = 0;

  for (let i = 0; i < ITERATIONS; i++) {
    const startForward = performance.now();
    const [z1, a1, z2, probs] = forward(X, W1, b1, W2, b2);
    forwardTime += performance.now() - startForward;

    const startBackward = performance.now();
    const [dW1, db1, dW2, db2] = backward(X, y, probs, z1, a1, W1, W2);
    backwardTime += performance.now() - startBackward;
  }

  console.log('Размеры матриц:');
  console.log(`Батч: ${BATCH_SIZE}`);
  console.log(`Вход: ${INPUT_SIZE}`);
  console.log(`Скрытый слой: ${HIDDEN_SIZE}`);
  console.log(`Выход: ${OUTPUT_SIZE}`);
  console.log(`\nВремя выполнения (${ITERATIONS} итераций):`);
  console.log(`Forward pass: ${(forwardTime / ITERATIONS / 1000).toFixed(6)} сек`);
  console.log(`Backward pass: ${(backwardTime / ITERATIONS / 1000).toFixed(6)} сек`);
  console.log(`Общее время: ${((forwardTime + backwardTime) / ITERATIONS / 1000).toFixed(6)} сек`);
}

main();
